src/main/kotlin/com/hypto/iam/server/idp/CognitoProviderImpl.kt

```diff
@@ -166,9 +173,16 @@ class CognitoIdentityProviderImpl : IdentityProvider, KoinComponent {
         val getUserRequest = AdminGetUserRequest.builder().userPoolId(identityGroup.id).username(userName).build()
         try {
             val response = cognitoClient.adminGetUser(getUserRequest)
+
+            // Adding this check as admin-get-user will succeed even with preferred_username
+            if (response.username() != userName) {
+                throw UserNotFoundException("User not found")
+            }
```

Comment:
As discussed, we should support getUser based on preferred userName. Since preferred username can be changed, we should ensure that before evaluating policy the preferred username gets resolved to the auto-generated username.

Also all places before we evaluate the policy, lets use auto-generated username instead of preferred username.

Reply:
For creating `ResourceHrn` for checking permissions, we are using this function 
https://github.com/hwslabs/iam/blob/88ab08e37b294506614ab2c5f7ebf15572a97ad2/src/main/kotlin/com/hypto/iam/server/security/Authorization.kt#L145-L161
So for user based apis like `/organizations/{organization_id}/users/{user_id}`, we have to add a special case for creating `ResourceHrn`, so that the userId provided in the url should not be used directly rather, we would fetch the actual username from cognito and use that.